src/ce/runprgm.src

```diff
@@ -0,0 +1,399 @@
+	ld	hl,(iy + 12)		; callback
+	add	hl,de
+	or	a,a
+	sbc	hl,de
+	ld	de,stub.callback - reloc.offset
+	jr	nz,.has_callback
+	dec	de
+	ld	hl,((-(stub.callback - stub.error + 1)) shl 8) or $18	; jr stub.error
+.has_callback:
+	ex	de,hl
+	ld	(hl),de
+	ld	bc,(iy + 9)
+	ld	(push_stub.data_size - reloc.offset),bc
+	ld	hl,(_exit.sp)
+	or	a,a
+	sbc	hl,bc
+	ld	de,(iy + 6)
+	di				; disable to prevent moving sp
+	push	hl
+	dec	hl
+	dec	hl
+	dec	hl
+	ld	(_exit.sp),hl
+	pop	hl
+	ld	iy,ti.flags
+	ex	de,hl
+	push	hl
+	or	a,a
+	sbc	hl,hl
+	sbc	hl,bc
+	pop	hl
+	jr	z,.done_move
+	push	hl
+	or	a,a
+	sbc	hl,de
+	pop	hl
+	jr	c,.lddr
+	ldir
+	jr	.done_move
+.lddr:
+	add	hl,bc
+	dec	hl
+	ex	de,hl
+	add	hl,bc
+	dec	hl
+	ex	de,hl
+	lddr 
+.done_move:
+	ld	hl,(_exit.sp)
+	ld	de,0			; fake return
+.fakeret := $-3
+	ld	(hl),de
+	or	a,a
+	sbc	hl,hl
+	jq	___exithl		; should return to caller
+
+virtual at ti.cursorImage
+launch_asm_prgm:
+	call	delete_running_prgm
+	call	setup_prgm_state
+	call	push_stub
+	ld	de,(___prgm_to_run_size)
+	ld	hl,ti.userMem
+	ex	hl,de
+	call	ti.InsertMem
+	ld	hl,___prgm_to_run
+	call	find_prgm
+	jr	c,.fatal
+	inc	hl
+	inc	hl
+	ld	de,ti.userMem
+	ld	bc,(___prgm_to_run_size)
+	ld	(ti.asm_prgm_size),bc
+	ldir
+	ld	hl,___prgm_to_run
+	call	ti.Mov9ToOP1
+	jp	ti.userMem
+.fatal:
+	ld	de,(___prgm_to_run_size)
+	ld	hl,ti.userMem
+	call	ti.DelMem
+	jp	ti.ErrMemory
+
+launch_basic_prgm:
+	call	delete_running_prgm
+	call	setup_prgm_state
+	call	push_stub
+	set	ti.progExecuting,(iy + ti.newDispF)
+	set	ti.cmdExec,(iy + ti.cmdFlags)
+	set	ti.allowProgTokens,(iy + ti.newDispF)
+	ld	hl,___prgm_to_run
+	call	ti.Mov9ToOP1
+	jp	ti.ParseInp
+
+setup_prgm_state:
+	xor	a,a
+	ld	(ti.kbdGetKy),a
+	jp	ti.DisableAPD
+
+delete_running_prgm:
+	ld	iy,ti.flags
+	ld	hl,ti.userMem
+	ld	de,(ti.asm_prgm_size)
+	call	ti.DelMem
+	or	a,a
+	sbc	hl,hl
+	ld	(ti.asm_prgm_size),hl
+	ret
+
+push_stub:
+	pop	hl
+	ld	(.ret),hl
+	ld	hl,stub.prgm
+	call	find_prgm
+	ld	(.stub_prgm_size),de
+	ld	(.prev_sp),sp
+	ld	hl,-9
+	add	hl,sp
+	ld	sp,hl
+	ld	(.stub_prgm),hl
+	ex	de,hl
+	ld	hl,stub.prgm
+	ld	bc,9
+	ldir
+	ld	hl,-(stub.len - 9)
+	add	hl,sp
+	ld	sp,hl
+	ld	(.stub_normal),hl
+	push	hl
+	ld	bc,stub.error_handler - stub.normal
+	add	hl,bc
+	ld	(.stub_error_handler),hl
+	pop	hl
+	ex	de,hl
+	ld	hl,stub
+	ld	bc,stub.len
+	ldir
+	ld	hl,0
+.prev_sp := $-3
+	push	hl
+	ld	hl,0
+.data_size := $-3
+	push	hl
+	ld	hl,0
+.stub_prgm := $-3
+	push	hl
+	ld	hl,0
+.stub_prgm_size := $-3
+	push	hl
+	ld	hl,0
+.stub_error_handler := $-3
+	call	ti.PushErrorHandler
+	ld	hl,0
+.stub_normal := $-3
+	push	hl
+	jp	0
+.ret := $-3
+
+find_prgm:
+	call	ti.Mov9ToOP1
+	call	ti.ChkFindSym
+	ret	c
+	call	ti.ChkInRam
+	ex	de,hl
+	jr	z,.in_ram
+	ld	de,9
+	add	hl,de
+	ld	e,(hl)
+	add	hl,de
+	inc	hl
+.in_ram:
+	call	ti.LoadDEInd_s
+	or	a,a
+	ret
+
+stub:
+.normal:
+	ld	(ti.cursorImage),hl
+	ld	iy,ti.flags
+	call	ti.PopErrorHandler
+	xor	a,a
+	sbc	hl,hl
+	ld	(ti.cursorImage),hl
+.basic_c_smc := $-3
+	jr	.no_error
+.error_handler:
```

Comment:
Not sure what's wrong with this, since I've never actually used error handlers before, but it seems to continue to execute the program after an error is encountered and the program calling RunPrgm exits. For example,
```
0/0
1
```
gives an `ERR: DATA TYPE` on the 1.
It seems to give an error unless the error occurs on the last line of the program, like it does in the test.

Reply:
Fixed by saving the parser state before executing the basic program